File: src/helpers/editor/syntax/python.ts
import store from "../../../store/index";
import { SyntaxHighlight, Token } from "../../../store/interfaces";

function createToken(color: string, value: string): Token {
  let token: Token = {
    value: value,
    color: color,
  };
  return token;
}

// prettier-ignore
let PYTHON_KEYWORDS: Array<string> = [
  "and", "as", "assert", "break", "class", "continue",
  "def", "del", "elif", "else", "except", "False",
  "finally", "for", "from", "global", "if", "import",
  "in", "is", "lambda", "None", "nonlocal", "not",
  "or", "pass", "raise", "return", "True", "try",
  "while", "with", "yield", "async", "await", "self",
  "match", "case"
];

let syntaxHighlight: SyntaxHighlight | null = null;
let langTokenTypes: any | null = null;

function handleWhitespace(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  color: string
): number {
  const startPos = pos;
  while (pos < inputChars.length && /\s/.test(inputChars[pos])) {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleComment(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  while (pos < inputChars.length && inputChars[pos] !== "\n") {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleStringLiteral(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  color: string
): number {
  const startChar = inputChars[pos];
  const startPos = pos;
  pos++; // Move past the opening quote

  // Check for triple-quoted string
  const isTripleQuote =
    pos + 1 < inputChars.length &&
    inputChars[pos] === startChar &&
    inputChars[pos + 1] === startChar;
  if (isTripleQuote) {
    pos += 2; // Move past triple quotes
    while (pos + 2 < inputChars.length) {
      if (
        inputChars[pos] === startChar &&
        inputChars[pos + 1] === startChar &&
        inputChars[pos + 2] === startChar
      ) {
        pos += 3; // End triple quotes
        break;
      }
      pos++;
    }
  } else {
    // Single-line string
    while (pos < inputChars.length && inputChars[pos] !== startChar) {
      if (inputChars[pos] === "\\" && pos + 1 < inputChars.length) {
        pos += 2;
        continue;
      }
      pos++;
    }
    pos++; // Move past closing quote
  }

  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleNumber(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  while (pos < inputChars.length && /\d/.test(inputChars[pos])) {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleReturnOperator(tokens: Token[], pos: number, color: string): number {
  tokens.push(createToken(color, "->"));
  return pos + 2;
}

function handleOperator(tokens: Token[], currentChar: string, pos: number, color: string): number {
  tokens.push(createToken(color, currentChar));
  return pos + 1;
}

function handleIdentifier(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  functionNameColor: string,
  keywordColor: string,
  classNameColor: string,
  identifierColor: string
): number {
  const startPos = pos;
  while (pos < inputChars.length && (/\w/.test(inputChars[pos]) || inputChars[pos] === "_")) {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  const nextChar = inputChars[pos] || " ";

  if (nextChar === "(" || inputChars[startPos - 1] === ".") {
    tokens.push(createToken(functionNameColor, value));
  } else {
    if (PYTHON_KEYWORDS.includes(value)) {
      tokens.push(createToken(keywordColor, value));
    } else if (value[0] === value[0].toUpperCase()) {
      tokens.push(createToken(classNameColor, value));
    } else {
      tokens.push(createToken(identifierColor, value));
    }
  }

  return pos;
}

function handlePunctuation(
  tokens: Token[],
  currentChar: string,
  pos: number,
  color: string
): number {
  tokens.push(createToken(color, currentChar));
  return pos + 1;
}

function handleUnknown(tokens: Token[], currentChar: string, pos: number, color: string): number {
  tokens.push(createToken(color, currentChar));
  return pos + 1;
}

// prettier-ignore
export function tokenizePythonCode(input: string): Token[] | null {
  let tokens: Token[] = [];

  if (!syntaxHighlight || !langTokenTypes) {
    syntaxHighlight = store.state.settings.syntaxHighlight;
    if (!syntaxHighlight) return null;

    langTokenTypes = syntaxHighlight.languages.find((lang) => lang.language === "python");
    if (!langTokenTypes) return null;
  }
  let pos = 0;
  const inputChars = Array.from(input);

  while (pos < inputChars.length) {
    const currentChar = inputChars[pos];

    if (/\s/.test(currentChar)) {
      pos = handleWhitespace(tokens, inputChars, pos, langTokenTypes["whitespaceColor"]);
    } 
    else if (currentChar === "#") {
      pos = handleComment(tokens, inputChars, pos, langTokenTypes["commentColor"]);
    } 
    else if (currentChar === '"' || currentChar === "'") {
      pos = handleStringLiteral(tokens, inputChars, pos, langTokenTypes["stringLiteralColor"]);
    } 
    else if (/\d/.test(currentChar)) {
      pos = handleNumber(tokens, inputChars, pos, langTokenTypes["numberColor"]);
    } 
    else if (currentChar === "-" && inputChars[pos + 1] === ">") {
      pos = handleReturnOperator(tokens, pos, langTokenTypes["operatorColor"]);
    } 
    else if ("+-*/=%".includes(currentChar)) {
      pos = handleOperator(tokens, currentChar, pos, langTokenTypes["operatorColor"]);
    } 
    else if (/[a-zA-Z_]/.test(currentChar)) {
      pos = handleIdentifier(
        tokens,
        inputChars,
        pos,
        langTokenTypes["functionNameColor"],
        langTokenTypes["keywordColor"],
        langTokenTypes["classNameColor"],
        langTokenTypes["identifierColor"]
      );
    } 
    else if (",.:;()[]{}'\"".includes(currentChar)) {
      pos = handlePunctuation(tokens, currentChar, pos, langTokenTypes["punctuationColor"]);
    } 
    else {
      pos = handleUnknown(tokens, currentChar, pos, langTokenTypes["unknownColor"]);
    }
  }
  return tokens;
}

File: src-tauri/src/syntax_highlight/languages/python.rs
use crate::settings::Settings;
use crate::syntax_highlight::{Token, TokenType, create_token};

const PYTHON_KEYWORDS: [&str; 38] = [
    "and", "as", "assert", "break", "class", "continue", 
    "def", "del", "elif", "else", "except", "False", 
    "finally", "for", "from", "global", "if", "import", 
    "in", "is", "lambda", "None", "nonlocal", "not", 
    "or", "pass", "raise", "return", "True", "try", 
    "while", "with", "yield", "async", "await", "self", 
    "match", "case"
];

fn handle_whitespace(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos].is_whitespace() {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "python", TokenType::Whitespace, &value));
}

fn handle_comment(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos] != '\n' {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "python", TokenType::Comment, &value));
}

fn handle_string_literal(
    tokens: &mut Vec<Token>, 
    settings: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_char = input_chars[*pos];
    let start_pos = *pos;
    *pos += 1; // Move past the opening quote

    // Check if this is a triple-quoted string
    let is_triple_quote = *pos + 1 < input_chars.len() && input_chars[*pos] == start_char && input_chars[*pos + 1] == start_char;

    if is_triple_quote {
        *pos += 2; // Move past the opening triple quotes

        // Process the content inside the triple quotes
        while *pos + 2 < input_chars.len() {
            // Check if we've reached the closing triple quotes
            if input_chars[*pos] == start_char && input_chars[*pos + 1] == start_char && input_chars[*pos + 2] == start_char {
                *pos += 3; // Move past the closing triple quotes
                break;
            }
            *pos += 1;
        }
    } else {
        // Handle single-line strings (delimited by ' or ")
        while *pos < input_chars.len() && input_chars[*pos] != start_char {
            // Allow for escaped quotes within single-line strings
            if input_chars[*pos] == '\\' && *pos + 1 < input_chars.len() {
                *pos += 2;
                continue;
            }
            *pos += 1;
        }
        *pos += 1; // Move past the closing quote
    }

    // Collect the entire string value including quotes
    let value: String = input_chars[start_pos..*pos.min(&mut input_chars.len())].iter().collect();
    tokens.push(create_token(settings, "python", TokenType::StringLiteral, &value));
}


fn handle_number(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos].is_digit(10) {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "python", TokenType::Number, &value));
}

fn handle_return_operator(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    pos: &mut usize
) {
    let value = "->".to_string();
    tokens.push(create_token(settings_config, "python", TokenType::Operator, &value));
    *pos += 2;
}

fn handle_operator(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    pos: &mut usize
) {
    let value = current_char.to_string();
    tokens.push(create_token(settings_config, "python", TokenType::Operator, &value));
    *pos += 1;
}

fn handle_identifier(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && (input_chars[*pos].is_alphanumeric() || input_chars[*pos] == '_') {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    let next_char = input_chars.get(*pos).unwrap_or(&' ');

    if *next_char == '(' || input_chars.get(start_pos.wrapping_sub(1)) == Some(&'.') {
        tokens.push(create_token(settings_config, "python", TokenType::FunctionName, &value));
    } else {
        let token_type;
        if PYTHON_KEYWORDS.contains(&value.as_str()) {
            token_type = TokenType::Keyword;
        } else if value.chars().next().unwrap().is_uppercase() {
            token_type = TokenType::ClassName;
        } else {
            token_type = TokenType::Identifier;
        }
        tokens.push(create_token(settings_config, "python", token_type, &value));
    }
}

fn handle_punctuation(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    pos: &mut usize
) {
    let value = current_char.to_string();
    tokens.push(create_token(settings_config, "python", TokenType::Punctuation, &value));
    *pos += 1;
}

fn handle_unknown(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    pos: &mut usize
) {
    let value = current_char.to_string();
    tokens.push(create_token(settings_config, "python", TokenType::Unknown, &value));
    *pos += 1;
}

pub fn tokenize_python_code(settings_config: &Settings, input: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut pos = 0;
    let input_chars: Vec<char> = input.chars().collect();

    while pos < input_chars.len() {
        let current_char = input_chars[pos];

        if current_char.is_whitespace() {
            handle_whitespace(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '#' {
            handle_comment(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '"' || current_char == '\'' {
            handle_string_literal(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char.is_digit(10) {
            handle_number(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '-' && input_chars.get(pos + 1) == Some(&'>') {
            handle_return_operator(&mut tokens, settings_config, &mut pos);
        } 
        else if "+-*/=%".contains(current_char) {
            handle_operator(&mut tokens, settings_config, current_char, &mut pos);
        } 
        else if current_char.is_alphabetic() || current_char == '_' {
            handle_identifier(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if ",.:;()[]{}'\"".contains(current_char) {
            handle_punctuation(&mut tokens, settings_config, current_char, &mut pos);
        } 
        else {
            handle_unknown(&mut tokens, settings_config, current_char, &mut pos);
        }
    }

    tokens
}